### PR TITLE
PUBDEV-8947: aic glm

### PR DESCRIPTION
**Pull request by syzonyuliia**
https://h2oai.atlassian.net/browse/PUBDEV-8947

### CODE REVIEW

**Comment by wendycwong** [comment]
@syzonyuliia :  I have added a R unit test for you that attempts to compare likelihood/aic calculation for GLM.  You can do a 
git cherry-pick  d4708dfc501a93b2cfd4751e6c7b79801c327aab

**Review by wendycwong — CHANGES_REQUESTED** [review]

**Comment by wendycwong** [comment]
Something is wrong.  Here when I run the comparison with R, here is what I got:

image

File: h2o-r/tests/testdir_algos/glm/runit_PUBDEV_8947_GLM_likelihoods_aic.R [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+test.glm.aic.likelihood <- function() {
+  # checking for Gaussian
+  h2o.data = h2o.uploadFile(locate("smalldata/prostate/prostate_complete.csv.zip"), destination_frame="h2o.data")    
+  R.data = as.data.frame(as.matrix(h2o.data))
+  myY = "GLEASON"
+  myX = c("ID","AGE","RACE","CAPSULE","DCAPS","PSA","VOL","DPROS")
+  R.formula = (R.data[,"GLEASON"]~.) 
+  model.h2o.gaussian.identity <- h2o.glm(calc_like=TRUE, x=myX, y=myY, training_frame=h2o.data, family="gaussian", link="identity",alpha=0.5, lambda=0, nfolds=0)
+  model.R.gaussian.identity <- glm(formula=R.formula, data=R.data[,2:9], family=gaussian(link=identity), na.action=na.omit)
+  perf <- h2o.performance(model.h2o.gaussian.identity, h2o.data)
```
Comment (wendycwong) [inline h2o-r/tests/testdir_algos/glm/runit_PUBDEV_8947_GLM_likelihoods_aic.R:13]: change 

perf <- h2o.performance(model.h2o.gaussian.identy, h2o.data)

to

perf <- h2o.performance(model.h2o.gaussian.identy, train=TRUE)

File: h2o-r/tests/testdir_algos/glm/runit_PUBDEV_8947_GLM_likelihoods_aic.R [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+ # browser()
+  
+  # check for Gamma
+  myY = "DPROS"
+  myX = c("ID","AGE","RACE","CAPSULE","DCAPS","PSA","VOL","GLEASON")
+  R.formula = (R.data[,"DPROS"]~.) 
+  model.h2o.gamma.inverse <- h2o.glm(calc_like=TRUE, x=myX, y=myY, training_frame=h2o.data, family="gamma", link="inverse",alpha=0.5, lambda=0, nfolds=0)
+  model.R.gamma.inverse <- glm(formula=R.formula, data=R.data[,c(1:5,7:9)], family=Gamma(link=inverse), na.action=na.omit)
+  perf <- h2o.performance(model.h2o.gamma.inverse, h2o.data)
+  print("GLM Gamma")
```
Comment (wendycwong) [inline h2o-r/tests/testdir_algos/glm/runit_PUBDEV_8947_GLM_likelihoods_aic.R:32]: change

perf <- h2o.performance(model.h2o.gamma.inverse, h2o.data)

to

perf <- h2o.performance(model.h2o.gamma.inverse, train=TRUE)

File: h2o-r/tests/testdir_algos/glm/runit_PUBDEV_8947_GLM_likelihoods_aic.R [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  print(logLik(model.R.gamma.inverse))
+  
+  # checking for binomial
+  myY <- "CAPSULE"
+  myX <- c("AGE","RACE","DCAPS","PSA","VOL","DPROS","GLEASON")
+  R.formula <- (R.data[,"CAPSULE"]~.)
+  model.h2o.binomial.logit <- h2o.glm(calc_like=TRUE, x=myX, y=myY, training_frame=h2o.data, family="binomial", link="logit",alpha=0.5, lambda=0, nfolds=0)
+  model.R.binomial.logit <- glm(formula=R.formula, data=R.data[,4:10], family=binomial(link=logit), na.action=na.omit)
+  perf <- h2o.performance(model.h2o.binomial.logit, h2o.data)
+  print("GLM binomial")
```
Comment (wendycwong) [inline h2o-r/tests/testdir_algos/glm/runit_PUBDEV_8947_GLM_likelihoods_aic.R:49]: change

perf <- h2o.performance(model.h2o.binomial.logit, h2o.data)

to

perf <- h2o.performance(model.h2o.binomial.logit, train=TRUE)

File: h2o-r/tests/testdir_algos/glm/runit_PUBDEV_8947_GLM_likelihoods_aic.R [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  print(logLik(model.R.binomial.logit))
+  
+  # checking for Poisson
+  myY = "GLEASON"
+  myX = c("ID","AGE","RACE","CAPSULE","DCAPS","PSA","VOL","DPROS")
+  R.formula = (R.data[,"GLEASON"]~.) 
+  model.h2o.poisson.log <- h2o.glm(calc_like=TRUE, x=myX, y=myY, training_frame=h2o.data, family="poisson", link="log",alpha=0.5, lambda=0, nfolds=0)
+  model.R.poisson.log <- glm(formula=R.formula, data=R.data[,2:9], family=poisson(link=log), na.action=na.omit)
+  perf <- h2o.performance(model.h2o.poisson.log, h2o.data)
+  print("GLM Poisson")
```
Comment (wendycwong) [inline h2o-r/tests/testdir_algos/glm/runit_PUBDEV_8947_GLM_likelihoods_aic.R:66]: change

perf <- h2o.performance(model.h2o.poisson.log, h2o.data)

to

perf <- h2o.performance(model.h2o.poisson.log, train=TRUE)

**Review by wendycwong — CHANGES_REQUESTED** [review]

File: h2o-algos/src/main/java/hex/glm/GLMModel.java [hunk 5]
```diff
@@ -713,6 +723,64 @@ public final double likelihood(double yr, double ym){
...
+      double probabilityOf1;
+      switch (_family) {
+        case gaussian:
+          return -.5 * (w * Math.pow(yr - prediction , 2) / _dispersion_estimated 
+                  + log(_dispersion_estimated / w) + LOG2PI);
+        case binomial:
+          // if probability is not given, then it is 1.0 if 1 is predicted and 0.0 if 0 is predicted
+          probabilityOf1 = ym.length > 1 ? ym[2] : ym[0]; // probability of 1 equals prediction
+          return w * (yr * log(probabilityOf1) + (1-yr) * log(1 - probabilityOf1)) 
+                  + w * (Gamma.digamma(2) - Gamma.digamma(yr + 1)
```
Comment (wendycwong) [inline h2o-algos/src/main/java/hex/glm/GLMModel.java:738]: Replace

Gamma.digamma

with

Gamma.logGamma

**Review by wendycwong — CHANGES_REQUESTED** [review]

**Comment by wendycwong** [comment]
@syzonyuliia :  please do:

git cherry-pick  5d40342e1cd23ea852d88fb65edf0d26a5d13cee

I have changed the R test and added a Java test for you to manually check the likelihood results.  I only added the binomial for you.  Please add the remaining ones yourself.

File: h2o-algos/src/main/java/hex/glm/GLMModel.java [hunk 6]
```diff
@@ -718,6 +728,58 @@ public final double likelihood(double yr, double ym){
...
+            return -w * (yr * log(probabilityOf1) + (1 - yr) * log(1 - probabilityOf1));
+        case fractionalbinomial:
+          // if probability is not given, then it is 1.0 if 1 is predicted and 0.0 if 0 is predicted
+          probabilityOf1 = ym.length > 1 ? ym[2] : ym[0]; // probability of 1 equals prediction
+          if (yr == prediction)
+            return 0;
+          return w * ((MathUtils.y_log_y(yr, probabilityOf1)) + MathUtils.y_log_y(1 - yr, 1 - probabilityOf1));
+        case poisson:
+          return w * (yr * log(prediction) - prediction - Gamma.digamma(yr + 1)); // gamma(n) = (n-1)!
+        case negativebinomial:
```
Comment (wendycwong) [inline h2o-algos/src/main/java/hex/glm/GLMModel.java:762]: change Gamma.diagamma(yr+1) to Gamma.logGamma(yr+1)

File: h2o-algos/src/main/java/hex/glm/GLMModel.java [hunk 7]
```diff
@@ -718,6 +728,58 @@ public final double likelihood(double yr, double ym){
...
+          return yr * log(_invTheta * prediction / w) - (yr + w/_invTheta) * log(1 + _invTheta * prediction / w) 
+                  + log(Gamma.gamma(yr + w/_invTheta) / (Gamma.gamma(yr + 1) * Gamma.gamma(w/_invTheta)));
+        case gamma:
+          return w * log(w * yr / prediction) - w * yr / prediction - log(yr) - Gamma.digamma(w);
+        case tweedie:
```
Comment (wendycwong) [inline h2o-algos/src/main/java/hex/glm/GLMModel.java:767]: Please change Gamma.digamma(w) to Gamma.logGamma(w)

File: h2o-algos/src/main/java/hex/glm/GLMModel.java [hunk 8]
```diff
@@ -718,6 +728,58 @@ public final double likelihood(double yr, double ym){
...
+          if (yr == prediction)
+            return 0;
+          return w * ((MathUtils.y_log_y(yr, probabilityOf1)) + MathUtils.y_log_y(1 - yr, 1 - probabilityOf1));
+        case poisson:
+          return w * (yr * log(prediction) - prediction - Gamma.digamma(yr + 1)); // gamma(n) = (n-1)!
+        case negativebinomial:
+          return yr * log(_invTheta * prediction / w) - (yr + w/_invTheta) * log(1 + _invTheta * prediction / w) 
+                  + log(Gamma.gamma(yr + w/_invTheta) / (Gamma.gamma(yr + 1) * Gamma.gamma(w/_invTheta)));
+        case gamma:
+          return w * log(w * yr / prediction) - w * yr / prediction - log(yr) - Gamma.digamma(w);
```
Comment (wendycwong) [inline h2o-algos/src/main/java/hex/glm/GLMModel.java:766]: The gamma family needs dispersion parameter estimation as well.  Please put the dispersion parameter in here.  
image

File: h2o-algos/src/main/java/hex/glm/GLMModel.java [hunk 9]
```diff
@@ -718,6 +728,58 @@ public final double likelihood(double yr, double ym){
...
+          // if probability is not given, then it is 1.0 if 1 is predicted and 0.0 if 0 is predicted
+          probabilityOf1 = ym.length > 1 ? ym[2] : ym[0]; // probability of 1 equals prediction
+          if (yr == prediction)
+            return 0;
+          return w * ((MathUtils.y_log_y(yr, probabilityOf1)) + MathUtils.y_log_y(1 - yr, 1 - probabilityOf1));
+        case poisson:
+          return w * (yr * log(prediction) - prediction - Gamma.digamma(yr + 1)); // gamma(n) = (n-1)!
+        case negativebinomial:
+          return yr * log(_invTheta * prediction / w) - (yr + w/_invTheta) * log(1 + _invTheta * prediction / w) 
+                  + log(Gamma.gamma(yr + w/_invTheta) / (Gamma.gamma(yr + 1) * Gamma.gamma(w/_invTheta)));
```
Comment (wendycwong) [inline h2o-algos/src/main/java/hex/glm/GLMModel.java:764]: The _invTheta here is the dispersion parameter.  Please enable the dispersion parameter calculation here as well and use that.

File: h2o-algos/src/main/java/hex/schemas/GLMV3.java [hunk 10]
```diff
@@ -64,7 +64,7 @@ public static final class GLMParametersV3 extends ModelParametersSchemaV3<GLMPar
             "link",
             "rand_link", // link function for random components, array
             "startval",  // initial starting values for fixed and randomized coefficients, double array
-            "calc_like", // HGLM, true will return likelhood function value
+            "calc_like", // true will return likelihood function value
             "HGLM",  // boolean: true - enabled HGLM, false - normal GLM
```
Comment (wendycwong) [inline h2o-algos/src/main/java/hex/schemas/GLMV3.java:68]: Please change comment for calc_like here:

If not HGLM, true will return likelihood function value.  

Then in the parameter definition part, please make the change as well here around line 215 of the code in GLMV3.java
image

File: h2o-algos/src/test/java/hex/glm/GLMTestAICLikelihood.java [hunk 11]
```diff
@@ -0,0 +1,388 @@
...
+      double logLike = 0.0;
+      double probabilityOf1;
+      double w = 1.0;
+      double yr;
+      for (int index=0; index<nRow; index++) {
+        yr = responseCol.at(index);
+        probabilityOf1 = pred.numCols() > 1 ? pred.vec(2).at(index) : pred.vec(0).at(index); // probability of 1 equals prediction
+        logLike +=  w * (trainData.vec("CAPSULE").at(index) * log(probabilityOf1) + (1-yr) * log(1 - probabilityOf1))
+                + w * (Gamma.digamma(2) - Gamma.digamma(yr + 1)
+                - Gamma.digamma(1 - yr + 1));
```
Comment (wendycwong) [inline h2o-algos/src/test/java/hex/glm/GLMTestAICLikelihood.java:102]: Change Gamma.digamma() to Gamma.logGamma()

File: h2o-algos/src/test/java/hex/glm/GLMTestAICLikelihood.java [hunk 12]
```diff
@@ -0,0 +1,388 @@
...
+      //manually calculate loglikelihood and AIC directly from formula
+      int nRow = (int) pred.numRows();
+      double logLike = 0.0;
+      double prediction;
+      double yr;
+      for (int index=0; index<nRow; index++) {
+        yr = responseCol.at(index);
+        prediction = pred.vec(0).at(index);
+        if(!Double.isNaN(yr) && !Double.isNaN(prediction)) {
+          logLike += yr * log(prediction) - prediction - Gamma.digamma(yr + 1);
```
Comment (wendycwong) [inline h2o-algos/src/test/java/hex/glm/GLMTestAICLikelihood.java:227]: change Gamma.digamma to Gamma.logGamma

File: h2o-r/h2o-package/R/models.R [hunk 13]
```diff
@@ -2044,6 +2044,69 @@ h2o.aic <- function(object, train=FALSE, valid=FALSE, xval=FALSE) {
...
+#' h2o.init()
+#' prostate_path <- system.file("extdata", "prostate.csv", package = "h2o")
+#' prostate <- h2o.uploadFile(path = prostate_path)
+#' p_sid <- h2o.runif(prostate)
+#' prostate_train <- prostate[p_sid > .2,]
+#' prostate_glm <- h2o.glm(x = 3:7, y = 2, training_frame = prostate_train)
+#' ll_basic <- h2o.loglikelihood(prostate_glm)
+#' print(ll_basic)
+#' }
+#' @export
```
Comment (wendycwong) [inline h2o-r/h2o-package/R/models.R:2072]: Please add a test with cross-validation on and with a validation frame and access the likelhood functions for training, cross-validation and valiation.

**Review by wendycwong — CHANGES_REQUESTED** [review]
Most are only small changes.

**Review by wendycwong — CHANGES_REQUESTED** [review]
Most are only small changes.

File: h2o-r/h2o-package/R/models.R [hunk 14]
```diff
@@ -2044,6 +2044,69 @@ h2o.aic <- function(object, train=FALSE, valid=FALSE, xval=FALSE) {
...
+h2o.loglikelihood <- function(object, train=FALSE, valid=FALSE, xval=FALSE) {
```
Comment (tomasfryda) [inline h2o-r/h2o-package/R/models.R:2073]: please update the `h2o-r/h2o-package/pkgdown/_pkgdown.yml` by adding this function to the reference.contents list

File: h2o-algos/src/main/java/hex/schemas/GLMV3.java (hunk 10 above)
Comment (syzonyuliia) [reply h2o-algos/src/main/java/hex/schemas/GLMV3.java:68]: For the line 67: This parameter is used by both HGLM and non-HGLM, so I just removed the limitation on the 67th line. Doe is make sense? In any case it controls the calculation of loglikelihood.

For the line 215: Thank you for pointing it out, I missed this comment. I would do it the same way as on the line 67 - "if true, will return likelihood function value." Is it OK for you?

File: h2o-algos/src/main/java/hex/glm/GLMModel.java [hunk 15]
```diff
@@ -718,6 +728,60 @@ public final double likelihood(double yr, double ym){
...
+          // if probability is not given, then it is 1.0 if 1 is predicted and 0.0 if 0 is predicted
+          probabilityOf1 = ym.length > 1 ? ym[2] : ym[0]; // probability of 1 equals prediction
+          if (yr == prediction)
+            return 0;
+          return w * ((MathUtils.y_log_y(yr, probabilityOf1)) + MathUtils.y_log_y(1 - yr, 1 - probabilityOf1));
+        case poisson:
+          return w * (yr * log(prediction) - prediction - Gamma.logGamma(yr + 1)); // gamma(n) = (n-1)!
+        case negativebinomial:
+          double invThetaEstimated = 1 / _dispersion_estimated;
+          return yr * log(invThetaEstimated * prediction / w) - (yr + w/invThetaEstimated) * log(1 + invThetaEstimated * prediction / w)
```
Comment (wendycwong) [inline h2o-algos/src/main/java/hex/glm/GLMModel.java:764]: Here is our formula for negative binomial:
image

Here is from outside doc:
image

I believe, our theta = k in the other document.  

So, please correct the formua as

yr * log(dispersion_estimated*prediction/w)-(yr+w*invThetaEstimated)*log(1+_dispersion_parameter*prediction/w)+Gamma.logGamma(yr+w*invThetaEstimated)-Gamma.logGamma(yr+1)-Gamma.logGamma(w*invThetaEstiamted)

**Review by wendycwong — CHANGES_REQUESTED** [review]

File: h2o-algos/src/main/java/hex/glm/GLM.java [hunk 16]
```diff
@@ -1281,6 +1281,28 @@ public void init(boolean expensive) {
         if (_parms._fix_tweedie_variance_power && !_parms._fix_dispersion_parameter)
           _tweedieDispersionOnly = true;
       
+        // likelihood calculation for gaussian, gamma, negativebinomial and tweedie families requires dispersion parameter estimation
+        // _dispersion_parameter_method: gaussian - pearson; gamma, negativebinomial, tweedie - ml.
+        if(!_parms._HGLM && _parms._calc_like) {
+          switch (_parms._family) {
+            case gaussian:
+              _parms._dispersion_parameter_method = pearson;
```
[the remark this answers is not in the document]
Comment (syzonyuliia) [reply h2o-algos/src/main/java/hex/glm/GLM.java:1289]: Your suggestion does not allow to set the ML to gamma and negativebinomial. The gaussian has to be separated with break

File: h2o-algos/src/main/java/hex/glm/GLMModel.java [hunk 17]
```diff
@@ -718,6 +728,60 @@ public final double likelihood(double yr, double ym){
...
+      double probabilityOf1;
+      switch (_family) {
+        case gaussian:
+          return -.5 * (w * Math.pow(yr - prediction , 2) / _dispersion_estimated 
+                  + log(_dispersion_estimated / w) + LOG2PI);
+        case binomial:
+          // if probability is not given, then it is 1.0 if 1 is predicted and 0.0 if 0 is predicted
+          probabilityOf1 = ym.length > 1 ? ym[2] : ym[0]; // probability of 1 equals prediction
+          return w * (yr * log(probabilityOf1) + (1-yr) * log(1 - probabilityOf1)) 
+                  + w * (Gamma.logGamma(2) - Gamma.logGamma(yr + 1)
```
[the remark this answers is not in the document]
Comment (syzonyuliia) [reply h2o-algos/src/main/java/hex/glm/GLMModel.java:743]: You are right, thank you for this remark

File: h2o-algos/src/main/java/hex/glm/GLM.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-algos/src/main/java/hex/glm/GLM.java]: Good point.

File: h2o-algos/src/main/java/hex/glm/GLMModel.java [hunk 18]
```diff
@@ -718,6 +728,59 @@ public final double likelihood(double yr, double ym){
...
+          double invThetaEstimated = 1 / _dispersion_estimated;
+          return yr * log(invThetaEstimated * prediction / w) - (yr + w/invThetaEstimated) * log(1 + invThetaEstimated * prediction / w) 
+                  + log(Gamma.gamma(yr + w / invThetaEstimated) / (Gamma.gamma(yr + 1) * Gamma.gamma(w / invThetaEstimated)));
+        case gamma:
+          double invPhiEst = 1 / _dispersion_estimated;
+          return w * invPhiEst * log(w * yr * invPhiEst / prediction) - w * yr * invPhiEst / prediction 
+                  - log(yr) - Gamma.logGamma(w * invPhiEst);
+        case tweedie:
+          if (DispersionMethod.ml.equals(_dispersion_parameter_method) && !_fix_tweedie_variance_power) {
+            return -TweedieVariancePowerMLEstimator.logLikelihood(yr, ym[0], _tweedie_variance_power, _init_dispersion_parameter);
```
Comment (wendycwong) [inline h2o-algos/src/main/java/hex/glm/GLMModel.java:771]: Change 

if (DispersionMethod.ml.equals(_dispersion_parameter_method) && !_fix_tweedie_variance_power)

to just 

if (DispersionMethod.ml.equals(_dispersion_parameter_method))

We should not care what the user sets when they want to estimate Tweedie dispersion parameter.  They should be able to get their likelihood calculation regardless.

**Review by wendycwong — APPROVED** [review]

**Review by wendycwong — APPROVED** [review]

File: h2o-algos/src/main/java/hex/glm/GLMModel.java (hunk 18 above)
Comment (syzonyuliia) [reply h2o-algos/src/main/java/hex/glm/GLMModel.java:771]: Oh right, and we also explicitly set the _dispersion_parameter_method to ML, so there is no reason to check for it

**Pull request merged**